### PR TITLE
fastn update

### PR DESCRIPTION
**Pull request by harshdoesdev**

### CODE REVIEW

File: fastn-update/src/lib.rs [hunk 1]
```diff
@@ -16,6 +16,12 @@ pub enum ManifestError {
         #[source]
         source: serde_json::Error,
     },
+    #[error("Failed to read manifest content for package '{package}'")]
+    DsReadError {
```
Comment (amitu) [inline fastn-update/src/lib.rs:20]: Rename this to `ReadError`.

File: fastn-update/src/lib.rs [hunk 2]
```diff
@@ -16,6 +16,12 @@ pub enum ManifestError {
...
+        package: String,
```
Comment (amitu) [inline fastn-update/src/lib.rs:21]: This is a bad pattern. Store the `package` in the error that stores `ManifestError`.

Comment (amitu) [reply fastn-update/src/lib.rs:21]: If you do this you can not use `#[from]` and create auto converters.

Comment (amitu) [reply fastn-update/src/lib.rs:21]: snafu solves this better: https://docs.rs/snafu/latest/snafu/trait.ResultExt.html#tymethod.context

File: fastn-core/tests/02-hello/cmd.p1 [hunk 3]
```diff
@@ -1,9 +1,11 @@
 -- fbt:
-cmd: cd amitu && $FBT_CWD/../target/debug/fastn --test build --edition 2022 --ignore-failed
+cmd: cd amitu && $FBT_CWD/../target/debug/fastn --test build --edition 2022 --ignore-failed --zip-url=https://codeload.github.com/amitu/dotcom/zip/refs/heads/main
 output: amitu/.build
 
 -- stdout:
 
+Processing www.amitu.com/manifest.json
+done in <omitted>
```
Comment (amitu) [inline fastn-core/tests/02-hello/cmd.p1:8]: Why is this two lines?

File: fastn-core/tests/02-hello/output/manifest.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    },
+    "index.ftd": {
+      "name": "index.ftd",
+      "hash": "0A0BF8A949025BA3F05B9D86F4302C9CCC90240E5B542584CDFCB1582677B760",
+      "size": 1045
+    }
+  },
+  "zip_url": "https://codeload.github.com/amitu/dotcom/zip/refs/heads/main",
+  "checksum": "F4CB85EF9F503C286A5C1FA656F60E849496B5EFA483AD86EB9942A0419675A6"
+}
```
Comment (amitu) [inline fastn-core/tests/02-hello/output/manifest.json:21]: Lets add a new line at then of file.

File: fastn-core/tests/03-nested-document/cmd.p1 [hunk 5]
```diff
@@ -1,10 +1,12 @@
 -- fbt:
-cmd: cd amitu && $FBT_CWD/../target/debug/fastn --test build --edition 2022
+cmd: cd amitu && $FBT_CWD/../target/debug/fastn --test build --edition 2022 --zip-url=https://codeload.github.com/amitu/dotcom/zip/refs/heads/main
```
Comment (amitu) [inline fastn-core/tests/03-nested-document/cmd.p1:2]: No, this information should be in FASTN.ftd file not here.

File: fastn-core/src/manifest/mod.rs [hunk 6]
```diff
@@ -53,17 +53,17 @@ pub async fn write_manifest_file(
...
-                        message: format!(
-                            "Could not find zip url for package \"{}\".",
-                            &config.package.name
-                        ),
-                    })
+                    fastn_core::warning!(
+                        "Could not find zip url for package \"{}\".",
+                        &config.package.name
+                    );
+
```
Comment (amitu) [inline fastn-core/src/manifest/mod.rs:65]: This is not a warning. Would everything work if this happens? If not let user know, what's the point of returning success and failing?

**Comment by amitu** [comment]
"Don't redownload a zip file if corresponding folder exists" is implemented @harshdoesdev?

**Comment by harshdoesdev** [comment]
> "Don't redownload a zip file if corresponding folder exists" is implemented @harshdoesdev?

Yes

File: fastn-core/src/manifest/mod.rs (hunk 6 above)
Comment (amitu) [reply fastn-core/src/manifest/mod.rs:65]: @harshdoesdev this is pending.

File: fastn-core/tests/03-nested-document/cmd.p1 (hunk 5 above)
Comment (amitu) [reply fastn-core/tests/03-nested-document/cmd.p1]: We have to remove `zip-url` as a cli flag.

File: fastn-core/src/manifest/mod.rs [hunk 7]
```diff
@@ -95,11 +95,16 @@ pub async fn write_manifest_file(
 
     let manifest = fastn_core::Manifest::new(files, zip_url, checksum);
 
+    let serialized_manifest = serde_json::ser::to_vec_pretty(&manifest)?;
+    let mut content_with_newline = serialized_manifest;
+    // Append newline character
+    content_with_newline.push(b'\n');
+
     let _ = &config
```
Comment (amitu) [inline fastn-core/src/manifest/mod.rs:102]: Delete `let _ =`

File: fastn-update/src/lib.rs [hunk 8]
```diff
@@ -249,6 +249,8 @@ async fn get_manifest(
     .context(DownloadManifestSnafu {
         package: package_name,
     })?;
+    // Append newline character
+    manifest_bytes.push(b'\n');
```
Comment (amitu) [inline fastn-update/src/lib.rs:253]: No.

Comment (amitu) [reply fastn-update/src/lib.rs:253]: DO NOT accept garbage. Do not fix other peoples garbage. Call things out. This is defensive programming, "oh I know they were supposed to generate the manifest file with new line in the end, but they screwed up, let me fix it for them". It you on both sides. You are writing code to generate manifest.json and you are writing code for downloading it. If you are generating manifest.json, do you have to add new line after downloading it?

File: fastn/src/main.rs [hunk 9]
```diff
@@ -246,7 +251,7 @@ fn app(version: &'static str) -> clap::Command {
                 .about("Build static site from this fastn package")
                 .arg(clap::arg!(file: [FILE]... "The file to build (if specified only these are built, else entire package is built)"))
                 .arg(clap::arg!(-b --base [BASE] "The base path.").default_value("/"))
-                .arg(clap::arg!(--"zip-url" ["ZIP URL"] "The zip url of the package."))
+                .arg(clap::arg!(--"zip-url" <URL> "The zip archive url for this package"))
```
Comment (amitu) [inline fastn/src/main.rs:254]: Let's delete this completely.

**Review by amitu — APPROVED** [review]

**Pull request merged**